R/external-generic.R

```diff
@@ -16,43 +16,90 @@
 #' when the other package is loaded.
 #'
 #' @param package Package the generic is defined in.
-#' @param generic Name of generic, as a string.
+#' @param name Name of generic, as a string.
 #' @param version An optional version the package must meet for the method to
 #'   be registered.
 #' @export
-new_external_generic <- function(package, generic, version = NULL) {
+new_external_generic <- function(package, name, version = NULL) {
   out <- list(
     package = package,
-    generic = generic,
+    name = name,
     version = version
   )
 
   class(out) <- "R7_external_generic"
   out
 }
 
+#' @export
+print.R7_external_generic <- function(x, ...) {
+  cat(
+    "<R7_external_generic> ",
+    x$package, "::", x$name, "()",
+    if (!is.null(x$version)) paste0(" (>= ", x$version, ")"),
+    "\n",
+    sep = ""
+  )
+  invisible(x)
+}
+
+is_external_generic <- function(x) {
+  inherits(x, "R7_external_generic")
+}
+
 #' @importFrom utils getFromNamespace packageName
 #' @rdname new_external_generic
+#' @param package Package name. Advanced use only.
 #' @export
 method_register <- function() {
```

Comment:
Yeah, and it's super confusing to have `register_method()` and `method_register()`. 

I'm not sure `external_methods_register()` is quite the right name because it feels like it's the generic that's external, not the methods, but it's definitely an improvement and we can iterate again in the future.